src/hw/raspi4/uart.c

```diff
@@ -0,0 +1,76 @@
+#include "port/raspi4/gpio.h"
+#include "port/raspi4/mbox.h"
+
+/* PL011 UART registers */
+#define UART0_DR   ((volatile unsigned int*)(MMIO_BASE + 0x00201000))
+#define UART0_FR   ((volatile unsigned int*)(MMIO_BASE + 0x00201018))
+#define UART0_IBRD ((volatile unsigned int*)(MMIO_BASE + 0x00201024))
+#define UART0_FBRD ((volatile unsigned int*)(MMIO_BASE + 0x00201028))
+#define UART0_LCRH ((volatile unsigned int*)(MMIO_BASE + 0x0020102C))
+#define UART0_CR   ((volatile unsigned int*)(MMIO_BASE + 0x00201030))
+#define UART0_IMSC ((volatile unsigned int*)(MMIO_BASE + 0x00201038))
+#define UART0_ICR  ((volatile unsigned int*)(MMIO_BASE + 0x00201044))
+
+void uart_init() {
+  register unsigned int r;
+
+  /* initialize UART */
+  *UART0_CR = 0; // turn off UART0
+
+  /* set up clock for consistent divisor values */
+  mbox[0] = 9 * 4;
```

Comment:
Please comment all these magic numbers with their origin.

E.g. (making this up but...)

mbox[0] = 9 * 4; // 9 whatevers and 4 magic unicorns

You the idea :)